src/librustc_mir/borrow_check/mod.rs

```diff
@@ -1808,15 +1724,162 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
             Read(ReadKind::Borrow(BorrowKind::Unique))
             | Read(ReadKind::Borrow(BorrowKind::Mut { .. }))
             | Read(ReadKind::Borrow(BorrowKind::Shared))
-            | Read(ReadKind::Copy) => {} // Access authorized
+            | Read(ReadKind::Copy) => { // Access authorized
+                return false;
+            }
         }
 
-        error_reported
+
+        // at this point, we have set up the error reporting state.
+
+        let mut err;
+        let note_the_place_err;
+        let item_msg = match self.describe_place(place) {
+            Some(name) => format!("immutable item `{}`", name),
+            None => "immutable item".to_owned(),
+        };
+
+        {
+            let local_info = if let Place::Local(local) = the_place_err {
+                Some(&self.mir.local_decls[*local])
+            } else {
+                None
+            };
+            debug!("check_access_permissions: \
+                    place {:?} is {} and the {:?} is erroneous \
+                    because of the_place_err {:?} local_info: {:?}",
+                   place, item_msg, error_access, the_place_err, local_info);
+        }
+
+        match error_access {
+            ErrorAccessKind::Mutate => {
+                let item_msg = match the_place_err {
+                    Place::Projection(box Projection {
+                        base: _,
+                        elem: ProjectionElem::Deref }
+                    ) => match self.describe_place(place) {
+                        Some(description) => format!("`{}`", description),
+                        None => format!("data in a `&` reference"),
+                    },
+                    _ => item_msg,
+                };
+                err = self.tcx.cannot_assign(span, &item_msg, Origin::Mir);
+            }
+            ErrorAccessKind::MutableBorrow => {
+                err = self.tcx
+                    .cannot_borrow_path_as_mutable(span, &item_msg, Origin::Mir);
+                err.span_label(span, "cannot borrow as mutable");
+            }
+        }
+
+        // SIGH: At this point, the invariant is that if error_access is MutableBorrow,
+        // then we've attached the main span_label and otherwise, we haven't.
+        //
+        // * It would be nicer to attach the span_label above in both
+        //   such cases, but doing so and keeping our diagnostics just
+        //   as informative would require even uglier contortions than
+        //   this one.
+
+        match (error_access, the_place_err) {
+            // We want to suggest users use `let mut` for local (user
+            // variable) mutations...
+            (_, Place::Local(local)) if {
+                // ... but it doesn't make sense to suggest it on
+                // variables that are already `ref x` or `ref mut x`
+                // (such variables are simply not mutable).
+                if let Some(ty::BindingMode::BindByValue(_)) =
```

Comment:
In such cases, often the user should have written `&mut *foo` instead of `&mut foo` — maybe we can think how to suggest that. (I've seen this before...)